Milestag2

vitos1k opened the pull request:
I've implemented MilesTag2 Lasertag protocol described here:
http://hosting.cmalton.me.uk/chrism/lasertag/MT2Proto.pdf

decoding  actuall meaning of bit's in protocol is on users side, they could just macro it  for example:
```cpp
#define MT_SHOT(player, team, damage) (unsigned long)(((player & 127) << 6) | ((team   & 3  ) << 4) | (damage & 15 ))
#define MT_PLAYER_VALUE(shot_data) (unsigned int)((shot_data >> 6) & 127)
#define MT_TEAM_VALUE(shot_data)   (unsigned short)((shot_data >> 4) & 3)
#define MT_DAMAGE_VALUE(shot_data) (unsigned short)(shot_data & 15)
```

To detect last bits in protocol, i've implemented functions with "space first, mark second" order, based on genericSend/matchGeneric

Tested it on ESP8266 and ESP32 and it works fine for me. But i don't have actual commercial Lasertag taggers and recievers with MilesTag2, to test against real world conditions. But i think this should be a good start.

X-Ref #1360

vitos1k committed:
Adding initial support for MilesTag2 protocol/testing github commits

vitos1k committed:
Added support for MilesTag protocol Sending and Decoding(recieving).
Added 2 decode types. 1 for 14bit Shooting packets. 2 for 24bit Message packets
According to this page https://github.com/crankyoldgit/IRremoteESP8266/wiki/Adding-support-for-a-new-IR-protocol

vitos1k committed:
Added some constants
Implementing more obvious and direct SEND function, to trace some SPACE inconsistency.

vitos1k committed:
MilesTag now have own function for sending and encoding, because of "Space" first, "Mark" second order

vitos1k committed:
My commit

- Bugfixing some intervals and found out that DutyCycle = 25-30% works great, even with my crappy vs1838b

- Implemented in decode GAP recognition in case of repeated message, but commented it out, because actual RAW GAP isn't 100ms as i set it, it actually is ~34ms, which is wierd, and i can't find out why

vitos1k committed:
Added forgotten newline at the end of file. Sorry

vitos1k committed:
- travis-ci autobuilding system was failing, on some constants borrowed from SonyProtocol, they were excessive. Fixed some assignements

vitos1k committed:
Tuned some repeat constants. Basically there is no documented information about REPEATs in MilesTag2 description, so they are arbitrary for now.

crankyoldgit commented:
I'll have a proper look through the PR soon, but here are some automated reported linter issues so far:
```
$ python cpplint.py --extensions=c,cc,cpp,ino --headers=h,hpp {src,src/locale,test,tools}/*.{h,c,cc,cpp,hpp,ino} examples/*/*.{h,c,cc,cpp,hpp,ino}

src/IRrecv.h:514:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/IRsend.h:240:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/IRsend.h:686:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/IRsend.h:687:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:9:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:9:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
src/ir_MilesTag2.cpp:9:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_MilesTag2.cpp:17:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:38:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
src/ir_MilesTag2.cpp:39:  Missing space after ,  [whitespace/comma] [3]
src/ir_MilesTag2.cpp:44:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
src/ir_MilesTag2.cpp:45:  Missing space after ,  [whitespace/comma] [3]
src/ir_MilesTag2.cpp:53:  Redundant blank line at the start of a code block should be deleted.  [whitespace/blank_line] [2]
src/ir_MilesTag2.cpp:56:  Should have a space between // and comment  [whitespace/comments] [4]
src/ir_MilesTag2.cpp:60:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:65:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:66:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:99:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:104:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:114:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:117:  Missing space after ;  [whitespace/semicolon] [3]
src/ir_MilesTag2.cpp:119:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:120:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:122:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:123:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
src/ir_MilesTag2.cpp:125:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:128:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_MilesTag2.cpp:133:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
src/ir_MilesTag2.cpp:142:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/IRsend.cpp:664:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/IRsend.cpp:718:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/IRsend.cpp:906:  Missing space after ,  [whitespace/comma] [3]
src/IRsend.cpp:909:  Missing space after ,  [whitespace/comma] [3]
src/IRsend.cpp:911:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/IRsend.cpp:911:  Should have a space between // and comment  [whitespace/comments] [4]
Total errors found: 35
```

and:
You need to add your new protocol name to `strToDecodeType()` and/or `typeToString()` in `IRutils.cpp`.
```
[ RUN      ] TestUtils.TypeStringConversionRangeTests
IRutils_test.cpp:501: Failure
Expected equality of these values:
  i
    Which is: 98
  strToDecodeType(typeToString((decode_type_t)i).c_str())
    Which is: -1
Protocol  doesn't decode from a string correctly
```

and some how you've modified `IRsend::defaultBits(decode_type_t::MIDEA24);` to give the incorrect value.
```
[ RUN      ] TestUtils.Housekeeping
ir_Midea_test.cpp:957: Failure
Expected equality of these values:
  kMidea24Bits
    Which is: 24
  IRsend::defaultBits(decode_type_t::MIDEA24)
    Which is: 14
[  FAILED  ] TestUtils.Housekeeping (0 ms)
```

and add a "Suported Devices:" comment to your file.
```
The following files had no supports section:
	ir_MilesTag2
```

vitos1k committed:
addressed all issues reported here https://github.com/crankyoldgit/IRremoteESP8266/pull/1380#issuecomment-757645909

vitos1k committed:
fixed own typo

vitos1k committed:
src/ir_MilesTag2.cpp:6 and 9:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]

vitos1k committed:
Fixed Support comment. But I don't have actual hardware on my hands right now

vitos1k commented:
> 
> 
> I'll have a proper look through the PR soon, but here are some automated reported linter issues so far:
> 
> ```
> $ python cpplint.py --extensions=c,cc,cpp,ino --headers=h,hpp {src,src/locale,test,tools}/*.{h,c,cc,cpp,hpp,ino} examples/*/*.{h,c,cc,cpp,hpp,ino}
> 
> src/IRrecv.h:514:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/IRsend.h:240:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/IRsend.h:686:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/IRsend.h:687:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:9:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:9:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
> src/ir_MilesTag2.cpp:9:  Should have a space between // and comment  [whitespace/comments] [4]
> src/ir_MilesTag2.cpp:17:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:38:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
> src/ir_MilesTag2.cpp:39:  Missing space after ,  [whitespace/comma] [3]
> src/ir_MilesTag2.cpp:44:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
> src/ir_MilesTag2.cpp:45:  Missing space after ,  [whitespace/comma] [3]
> src/ir_MilesTag2.cpp:53:  Redundant blank line at the start of a code block should be deleted.  [whitespace/blank_line] [2]
> src/ir_MilesTag2.cpp:56:  Should have a space between // and comment  [whitespace/comments] [4]
> src/ir_MilesTag2.cpp:60:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:65:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:66:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:99:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:104:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:114:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:117:  Missing space after ;  [whitespace/semicolon] [3]
> src/ir_MilesTag2.cpp:119:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:120:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:122:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:123:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
> src/ir_MilesTag2.cpp:125:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:128:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/ir_MilesTag2.cpp:133:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
> src/ir_MilesTag2.cpp:142:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/IRsend.cpp:664:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/IRsend.cpp:718:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/IRsend.cpp:906:  Missing space after ,  [whitespace/comma] [3]
> src/IRsend.cpp:909:  Missing space after ,  [whitespace/comma] [3]
> src/IRsend.cpp:911:  At least two spaces is best between code and comments  [whitespace/comments] [2]
> src/IRsend.cpp:911:  Should have a space between // and comment  [whitespace/comments] [4]
> Total errors found: 35
> ```
> 
> and:
> You need to add your new protocol name to `strToDecodeType()` and/or `typeToString()` in `IRutils.cpp`.
> 
> ```
> [ RUN      ] TestUtils.TypeStringConversionRangeTests
> IRutils_test.cpp:501: Failure
> Expected equality of these values:
>   i
>     Which is: 98
>   strToDecodeType(typeToString((decode_type_t)i).c_str())
>     Which is: -1
> Protocol  doesn't decode from a string correctly
> ```
> 
> and some how you've modified `IRsend::defaultBits(decode_type_t::MIDEA24);` to give the incorrect value.
> 
> ```
> [ RUN      ] TestUtils.Housekeeping
> ir_Midea_test.cpp:957: Failure
> Expected equality of these values:
>   kMidea24Bits
>     Which is: 24
>   IRsend::defaultBits(decode_type_t::MIDEA24)
>     Which is: 14
> [  FAILED  ] TestUtils.Housekeeping (0 ms)
> ```
> 
> and add a "Suported Devices:" comment to your file.
> 
> ```
> The following files had no supports section:
> 	ir_MilesTag2
> ```

Addressed most of the issues, except the support comment. I don't have actual Hardware (lasertag taggers and recievers) on my hands right now. And due to lockdown, i can't test it. I've added this comment 
```cpp
// Supports:
//   Brand: Theoretically,  Model: MILESTAG2 SUPPORTED HARDWARE
```

crankyoldgit commented:
Cool. Now that it's passing the automated checks, I'll try to get to reviewing it tomorrow.

crankyoldgit commented on src/IRrecv.h:
```diff
@@ -509,6 +509,11 @@ class IRrecv {
                       const bool strict = true);
 #endif
+#if DECODE_MILESTAG2
+  bool decodeMiles(decode_results *results, uint16_t offset = 0,
```
The default for `offset` should be `kStartOffset` aka `1`. Why is that not the case?

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -719,6 +719,15 @@
+#endif  // SEND_MILESTAG2
+
+
+
```
Remove two blank lines here.

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -1094,6 +1105,10 @@ const uint16_t kZepealBits = 16;
 const uint16_t kVoltasBits = 80;
 const uint16_t kVoltasStateLength = 10;
+const uint16_t kMilesTag14Bits = 14;
+const uint16_t kMilesTag24Bits = 24;
```
You should rename these to `kMilesTag2ShotBits` & `kMilesTage2MsgBits`  or something.

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -1094,6 +1105,10 @@ const uint16_t kZepealBits = 16;
+const uint16_t kMilesTag14Bits = 14;
+const uint16_t kMilesTag24Bits = 24;
+const uint16_t kMilesMinRepeat = 0;
+
```
Remove the extra blank line.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+/// @see http://hosting.cmalton.me.uk/chrism/lasertag/MT2Proto.pdf
+
+// Supports:
+//   Brand: Theoretically,  Model: MILESTAG2 SUPPORTED HARDWARE
```
I'd suggest the following instead:
```
//   Brand: Milestag2,  Model: Various
```

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+        mark(kMilesZeroMark);
+      }
+      space(kMilesRptLength);
+  }
```
You should be able to replace all of this with [sendGeneric](https://crankyoldgit.github.io/IRremoteESP8266/doxygen/html/classIRsend.html#a5215fd797dfd490816f31bb99b38c273):
e.g.
```c++
sendGeneric(
    kMilesHdrMark, kMilesSpace,  // Header
    kMilesOneMark, kMilesSpace,  // 1 bit
    kMilesZeroMark, kMilesSpace,  // 0 bit
    0,  // No footer mark
    kMilesRptLength, data, nbits, kMilesStdFreq, true,  // MSB First
    repeat, kMilesStdDuty);
```

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+  uint64_t data = 0;
+  // Header
+  if (!matchMark(*(results->rawbuf + offset++),
+                  kMilesHdrMark, kUseDefTol, kMarkExcess)) return 0;
```
FYI, I'm pretty sure this won't work as you expect.
The way your protocol seems to expect to decode is a header mark followed by either a one or a zero mark.
That would actually present itself as a single _longer_ header mark. i.e. a mark + another mark = just a single bigger mark.
There is typically a "space" between the two. The way your `_send` routine works there is a `kMilesSpace` between the two _marks_.

vitos1k replied:
The document clearly states, that after Header mark 2400us you definetly have a space 600us.
I just decided that it's easier to assume that "first space after header is space of first data bit", Instead of "first space is header space". That way the last bit would have "space and a (zero/one)mark" instead of "(zero/one)mark and silence", since space is 600us silence. Space first then goes a Mark of a bit. That's why i've rewrote genericSend and matchGeneric in my file.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+    } else {
+      return false;  // It's neither, so fail.
+    }
+  }
```
Assuming my other comment is correct, you can replace all/most of this with [matchGeneric](https://crankyoldgit.github.io/IRremoteESP8266/doxygen/html/classIRrecv.html#ab783f52acc2ff4052313d6947563e4fd):
e.g.
```c++
if (!matchGeneric(results->rawbuf + offset, &data,
                  results->rawlen - offset, nbits,
                  kMilesHdrMark, kMilesSpace,
                  kMilesOneMark, kMilesSpace,
                  kMilesZeroMark, kMilesSpace,
                  0, kMilesRptLength, true)) return false;
```

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+      break;
+    default:
+      return false;
+  }
```
I think this is a bad idea. The two different protocols are really just the same, but with different bit lengths.
e.g. Similar to `SONY`, which has three different bit lengths for the same Protocol (name).

Let's just make it a single protocol number/type of `decode_type_t::MILESTAG2`.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,155 @@
+}
+
+void IRsend::_sendMiles(const uint64_t data, const uint16_t nbits,
+                      const uint16_t repeat) {
```
Let's ditch the separate `send()` routines for msg & shot. Use a common `sendMilesTag2()` instead and let the user of that function choose what one they want by specifying the correct bit size.

crankyoldgit commented on src/locale/defaults.h:
```diff
@@ -628,6 +628,12 @@
+#endif  // D_STR_MILESTAG2SHOT
+#ifndef D_STR_MILESTAG2MSG
+#define D_STR_MILESTAG2MSG "MILESTAG2MSG"
+#endif  // D_STR_MILESTAG2MSG
```
See other comments/suggestions. Lets just go with a single protocol of `MILESTAG2`.

crankyoldgit commented on src/IRtext.cpp:
```diff
@@ -278,5 +278,7 @@ const PROGMEM char *kAllProtocolNamesStr =
     D_STR_ELITESCREENS "\x0"
     D_STR_PANASONIC_AC32 "\x0"
+    D_STR_MILESTAG2SHOT "\x0"
+    D_STR_MILESTAG2MSG "\x0"
```
See other comment. Lets move to a single protocol name.

crankyoldgit commented on src/IRsend.h:
```diff
@@ -682,6 +682,19 @@ class IRsend {
 #endif  // SEND_ELITESCREENS
+#if SEND_MILESTAG2
+  // Since There 2 types of transmissions
+  // (14bit for Shoting,24 bit for msg delivery) we use 2 separate functions
```
Typo: "Shooting, "

crankyoldgit commented on src/IRsend.cpp:
```diff
@@ -897,6 +901,14 @@ bool IRsend::send(const decode_type_t type, const uint64_t data,
+      case MILESTAG2MSG:
+        sendMilesMsg(data, nbits, min_repeat);
+        break;
+#endif  // SEND_MILESTAG2
```
Per other comments, lets just use a single common `sendMilestag2()` function rather that two separate ones as they are identical except for nr of bits.

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -867,8 +876,10 @@ enum decode_type_t {
   ELITESCREENS,  // 95
   PANASONIC_AC32,
+  MILESTAG2SHOT,
+  MILESTAG2MSG,
```
Make this only one protocol.
e.g.
`MILESTAG2,`

crankyoldgit commented on src/IRrecv.cpp:
```diff
@@ -983,6 +983,10 @@ bool IRrecv::decode(decode_results *results, irparams_t *save,
 #endif  // DECODE_PANASONIC_AC32
+#if DECODE_MILESTAG2
+    DPRINTLN("Attempting MilesTag2 decode");
+    if (decodeMiles(results, offset)) return true;
```
This needs to be:
```
    if (decodeMiles(results, offset, kMilesTag14Bits) || decodeMiles(results, offset, kMilesTag24Bits)) return true;
```

Otherwise you'll never detect/match the MSG messages.

crankyoldgit reviewed (dismissed):
Sorry for all the suggestions/corrections. You've done really well. Please don't see this as negative criticism. It's not meant to be.

I know you don't have access to any equipment etc at the moment, but do you have any recordings, dumps or other _code_ that we can use in the interim to confirm/check we are producing/matching a correct signal?

Once we get this sorted a bit more, I'll add some basic unit tests to this PR, to at least make sure we are internally consistent. i.e. We can send something and decode it back. It may not work with real devices, but that's better than than nothing! ;-)

vitos1k committed:
Removing some blank lines

Merging 2 functions and decode types into ONE Milestag2 as suggested by review

vitos1k committed:
extra space fix

vitos1k committed:
Replaced my functions with Built-in generic

vitos1k commented:
Addressed all the issues. But while testing it via **_IRrecvDumpV2.ino_**  i realized that Milestag2 protocol decoded as SONY protocol by library.
I send `irsend.sendMilestag2(0x379);`  
i also tested `irsend.sendSony38(0x379,14,0);` and got similar Dump
The only trouble, SONY woudn't decode 24bits  signal, in fact i'm surprised it even got decoded by SONY, since sony is working only with 12,15,20 bits messages.

```
 Timestamp : 000038.648
Library   : v2.7.13

Protocol  : SONY
Code      : 0x379 (14 Bits)
uint16_t rawData[29] = {2428, 612,  592, 592,  606, 602,  606, 600,  606, 600,  1210, 602,  1208, 602,  606, 602,  1208, 602,  1208, 600,  1210, 604,  1210, 602,  602, 604,  604, 604,  1214};  // SONY 379
uint64_t data = 0x379;


Timestamp : 000038.758
Library   : v2.7.13

Protocol  : SONY
Code      : 0x379 (14 Bits)
uint16_t rawData[29] = {2436, 606,  604, 602,  598, 608,  608, 598,  608, 598,  1210, 602,  1206, 604,  602, 604,  1212, 600,  1210, 600,  1210, 600,  1202, 610,  608, 600,  598, 584,  1236};  // SONY 379
uint64_t data = 0x379;
```

The only Issue i have now, is that ESP (irSend) is giving me bad timings right when i'm starting transmition. First ~5-8 bits (10-16 raw values) are OFF. And after that all timings become normal. I have to send it twice with short delay, and second time it goes alright. 
Here is dump of invalid Data:

```
Timestamp : 000377.323
Library   : v2.7.13

Protocol  : UNKNOWN
Code      : 0xD1D773FF (15 Bits)
uint16_t rawData[29] = {2630, 374,  834, 370,  832, 372,  812, 404,  786, 422,  1254, 556,  1268, 550,  646, 564,  1204, 604,  1210, 596,  1204, 598,  1206, 596,  614, 592,  608, 596,  1204};  // UNKNOWN D1D773FF


Timestamp : 000379.377
Library   : v2.7.13

Protocol  : UNKNOWN
Code      : 0xD1D773FF (15 Bits)
uint16_t rawData[29] = {2630, 374,  834, 370,  832, 372,  808, 406,  782, 424,  1254, 554,  1268, 546,  650, 566,  1204, 604,  1210, 594,  1206, 598,  1202, 602,  616, 590,  612, 590,  1202};  // UNKNOWN D1D773FF
```
Delay is ~2seconds between two messages. See how header is 200us bigger than it should, and header space is 200-300us shorter than it should be. I'm excpecting {2400, 600, 600, 600, 600, 600, 600, 600, 600, 600, 1200, 600ish values}
See how it becomes more or less better to the end of the message.
I've tested this on Arduino UNO as reciever too, and got similar Dump. 

My reciever is VS1838B, it's crappy, should i blame it?

vitos1k committed:
default offset is set to kStartOffset

crankyoldgit commented on src/IRrecv.h:
```diff
@@ -509,6 +509,11 @@ class IRrecv {
+#if DECODE_MILESTAG2
+  bool decodeMiles(decode_results *results, uint16_t offset = kStartOffset,
+                  const uint16_t nbits = kMilesTag2ShotBits,
+                  const bool strict = true);
```
Trivial style/readability nitpick: Please line up the `c` of `const` with the `d` of `decode_results`.

vitos1k replied:
done

crankyoldgit commented on src/IRsend.h:
```diff
@@ -682,6 +682,13 @@ class IRsend {
+  // (14bits for Shooting by default, you can set 24 bit for msg delivery)
+  void sendMilestag2(const uint64_t data,
+                 const uint16_t nbits = kMilesTag2ShotBits,
+                 const uint16_t repeat = kMilesMinRepeat);
```
Trivial style/readability nitpick: Please line up the `c`s of `const`s.

vitos1k replied:
done

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,105 @@
+
+// Constants
+// Constants are similar to Sony SIRC protocol, bassically they are very
+// similar, just nbits are varying
```
If I recall correctly, Sony's protocol also uses a different modulation frequency (40kHz vs 38kHz):
https://github.com/crankyoldgit/IRremoteESP8266/blob/8cf70043b7159e1ba672227fe13343cc3615e35e/src/ir_Sony.cpp#L35
and a different duty cycle 33% vs 25%:
https://github.com/crankyoldgit/IRremoteESP8266/blob/8cf70043b7159e1ba672227fe13343cc3615e35e/src/ir_Sony.cpp#L78
and Sony requires 3 messages (2 repeats).
https://github.com/crankyoldgit/IRremoteESP8266/blob/8cf70043b7159e1ba672227fe13343cc3615e35e/src/IRremoteESP8266.h#L1068

i.e. They differ more than just bit lengths.

crankyoldgit reviewed (commented):
> Addressed all the issues. But while testing it via **_IRrecvDumpV2.ino_** i realized that Milestag2 protocol decoded as SONY protocol by library.
> I send `irsend.sendMilestag2(0x379);`
> i also tested `irsend.sendSony38(0x379,14,0);` and got similar Dump
> The only trouble, SONY woudn't decode 24bits signal, in fact i'm surprised it even got decoded by SONY, since sony is working only with 12,15,20 bits messages.

Yeah, The code for decoding Sony is a tad of the "special" side. It was written ages ago, and to be as compact as possible, and match any/all of the Sony sizes (and then some) in an CPU efficient manor. i.e. It needs to be re-written.

You'll need to move the `decodeMilestag2()` etc code position in `IRrecv::decode()` to be before `decodeSony()` and add a note saying why. e.g. It's almost identical to SONY, but with different bit sizes. e.g. decodeSony() will match the 14 bit variant).
 
> 
> ```
>  Timestamp : 000038.648
> Library   : v2.7.13
> 
> Protocol  : SONY
> Code      : 0x379 (14 Bits)
> uint16_t rawData[29] = {2428, 612,  592, 592,  606, 602,  606, 600,  606, 600,  1210, 602,  1208, 602,  606, 602,  1208, 602,  1208, 600,  1210, 604,  1210, 602,  602, 604,  604, 604,  1214};  // SONY 379
> uint64_t data = 0x379;
> 
> 
> Timestamp : 000038.758
> Library   : v2.7.13
> 
> Protocol  : SONY
> Code      : 0x379 (14 Bits)
> uint16_t rawData[29] = {2436, 606,  604, 602,  598, 608,  608, 598,  608, 598,  1210, 602,  1206, 604,  602, 604,  1212, 600,  1210, 600,  1210, 600,  1202, 610,  608, 600,  598, 584,  1236};  // SONY 379
> uint64_t data = 0x379;
> ```

What you've listed here is a Sony msg, **and** one of it's repeats. That's why they are identical.
e.g. Timestamp `000038.758` - `000038.648` = 0.110 seconds.

> 
> The only Issue i have now, is that ESP (irSend) is giving me bad timings right when i'm starting transmition. First ~5-8 bits (10-16 raw values) are OFF. And after that all timings become normal. I have to send it twice with short delay, and second time it goes alright.
> Here is dump of invalid Data:
> 
> ```
> Timestamp : 000377.323
> Library   : v2.7.13
> 
> Protocol  : UNKNOWN
> Code      : 0xD1D773FF (15 Bits)
> uint16_t rawData[29] = {2630, 374,  834, 370,  832, 372,  812, 404,  786, 422,  1254, 556,  1268, 550,  646, 564,  1204, 604,  1210, 596,  1204, 598,  1206, 596,  614, 592,  608, 596,  1204};  // UNKNOWN D1D773FF
> 
> 
> Timestamp : 000379.377
> Library   : v2.7.13
> 
> Protocol  : UNKNOWN
> Code      : 0xD1D773FF (15 Bits)
> uint16_t rawData[29] = {2630, 374,  834, 370,  832, 372,  808, 406,  782, 424,  1254, 554,  1268, 546,  650, 566,  1204, 604,  1210, 594,  1206, 598,  1202, 602,  616, 590,  612, 590,  1202};  // UNKNOWN D1D773FF
> ```
> 
> Delay is ~2seconds between two messages. See how header is 200us bigger than it should, and header space is 200-300us shorter than it should be. I'm excpecting {2400, 600, 600, 600, 600, 600, 600, 600, 600, 600, 1200, 600ish values}
> See how it becomes more or less better to the end of the message.
> I've tested this on Arduino UNO as reciever too, and got similar Dump.
> 
> My reciever is VS1838B, it's crappy, should i blame it?

That's definitely peculiar. The code as it is now shouldn't produce that. (I'll add some unit tests to confirm that later)

The VS1838B could be to blame. Not sure. Does `sendSony()` and `sendSony38()` give similar results? If they do, then it's probably the VS1838B, if not, then it's likely to be the sendMilestag2()` code.

vitos1k committed:
decodeMiles() is set above decodeSony() to reduce false detection, since this two a similar

minor readability changes

vitos1k commented:
resolved all issues with readability. 


> That's definitely peculiar. The code as it is now shouldn't produce that. (I'll add some unit tests to confirm that later)
> 
> The VS1838B could be to blame. Not sure. Does `sendSony()` and `sendSony38()` give similar results? If they do, then it's probably the VS1838B, if not, then it's likely to be the sendMilestag2()` code.

sendSony38() gives exact same behaviour. So i guess it's my recievers problem. It looks like it needs some time to warmup at the start of each transmission.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+const uint16_t kMilesStdDuty = 25;
+
+
+
```
Remove two of these blank lines.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  if (results->rawlen>gap_pos) offset = gap_pos+1;
+  */ 
```
Please remove this commented out code.

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Victor Mukayev (vitos1k)
```
2021 ;-)

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+#include "IRsend.h"
+#include "IRutils.h"
+
+// Constants
```
Add constants:
```
const uint16_t kMilesShotMask = 1 << (kMilesTag2ShotBits - 1);   // Shot packets have this bit as `0`
const uint32_t kMilesMsgMask = 1 << (kMilesTag2MsgBits - 1);   // Msg packets  have this bit as `1`
const uint8_t  kMilesMsgTerminator = 0xE8;
```

Per http://hosting.cmalton.me.uk/chrism/lasertag/MT2Proto.pdf Section 2.1 & 2.2

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+  if (strict) {
+    switch (nbits) {  // Check we've been called with a correct bit size.
+      case 14:
+      case 24:
```
```
case kMilesTag2ShotBits:
case kMilesTag2MsgBits:
```

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+  results->value = data;
+  results->decode_type = decode_type_t::MILESTAG2;
+  results->command = 0;
+  results->address = 0;
```
remove. (Because it has been moved.)

crankyoldgit commented on src/ir_MilesTag2.cpp:
```diff
@@ -0,0 +1,102 @@
+                  kMilesOneMark, kMilesSpace,
+                  kMilesZeroMark, kMilesSpace,
+                  0, kMilesRptLength, true)) return false;
+  // Success
```
Add before `// Success`:
```
  results->command = 0;
  results->address = 0;

  // Compliance
  if (strict) {
    switch (nbits) {
      case kMilesTag2ShotBits:
        // Is it a valid shot packet?
        if (data & kMilesShotMask) return false;
        results->command = data & 0x3F;  // Team & Damage
        results->address = data >> 6;  // Player ID.
        break;
      case kMilesTag2MsgBits:
        // Is it a valid msg packet? i.e. Msg bit set & Terminator present.
        if (!(data & kMilesMsgMask) || ((data & 0xFF) != kMilesMsgTerminator))
          return false;
        results->command = (data >> 8) & 0xFF;  // Message data
        results->address = (data >> 16) & 0x7F;  // Message ID
        break;
    }
  }
```

Note: I haven't tested or compiled that code, so there could be errors.

crankyoldgit commented:
> resolved all issues with readability.
> 
> > That's definitely peculiar. The code as it is now shouldn't produce that. (I'll add some unit tests to confirm that later)
> > The VS1838B could be to blame. Not sure. Does `sendSony()` and `sendSony38()` give similar results? If they do, then it's probably the VS1838B, if not, then it's likely to be the sendMilestag2()` code.
> 
> sendSony38() gives exact same behaviour. So i guess it's my recievers problem. It looks like it needs some time to warmup at the start of each transmission.

IIRC, the VS1838B requires 5V, make sure you are supplying the voltage etc. i.e. Check your circuit. But yes, the VS1838B is not the _best_ demodulator out there.

It could also be your sending circuit too, or the ESP. But I'd wager it's the demodulator and/or circuit.

vitos1k committed:
As suggested, added MilesTag2 data decoding code https://github.com/crankyoldgit/IRremoteESP8266/pull/1380#discussion_r557357932

Minor readability fixes

vitos1k committed:
spaces in comments

vitos1k commented:
With `matchGeneric()` milesTag2 packets aren't decoded properly. Because as i mentioned when SENDING milestag2 signal,  last SPACE bit is just Silence (and there is no footer). And Recivers asume signal just stopped. And number of marks and spaces isn't even. Using DEBUG, i found that when calling `matchGeneric()` it goes okay, until last bit, it finds (one or zero mark) and then seeks for last space, but there is non, so it returns FAIL. 
That's why i used custom function, where i check for HEADER bit, but not HEADER_SPACE, instead i check regular SPACE, since it basically the same length as HEADER_SPACE,  and then a MARK. That way i end up with even number of SPACES and MARKS.

```cpp
  uint64_t data = 0;  
    // Header
  if (!matchMark(*(results->rawbuf + offset++),
                  kMilesHdrMark, kUseDefTol, kMarkExcess)) return 0;
  //  No need to check for HEADERSPACE since SPACE would be addressed imidiately in the next loop
  // Data
  uint16_t shift = 0;
  for (shift = 0; shift < nbits * 2; shift += 2) {
    // Is the bit a '1'?
    //  we check for SPACE first and MARK second, to accomodate uneven number of raw signals
    if (matchSpace(*(results->rawbuf + offset + shift),
                  kMilesSpace, kUseDefTol, kMarkExcess) &&
      matchMark(*(results->rawbuf + 1 + offset + shift),
                 kMilesOneMark, kUseDefTol, kMarkExcess))
         {
      data = (data << 1) | 1;
    } else if (matchSpace(*(results->rawbuf + offset + shift),
                         kMilesSpace, kUseDefTol, kMarkExcess) &&
               matchMark(*(results->rawbuf + 1 + offset + shift),
                        kMilesZeroMark, kUseDefTol, kMarkExcess))
                {
      data <<= 1;  // The bit is a '0'.
    } else {
      return false;  // It's neither, so fail.
    }
  }
```

Should i commit it? Tested it here, works fine

crankyoldgit commented:
> Should i commit it? Tested it here, works fine

No. This is a failure of `matchGeneric()`. I'll fix it so it works properly.

crankyoldgit committed:
Extend `matchData()` & `matchGeneric()` to handle no trailing space.
Note: No support yet for `matchBytes()`

crankyoldgit committed:
Add support for no trailing space for `matchBytes()`.

crankyoldgit committed:
Add basic unit tests
* Unit tests
  - Housekeeping
  - Sending Shot & Msg sizes.
  - Self-decoding
  - Failure when packets are not formatted correctly.
* Rename constants/functions to be consistent in style.
* Restructure `decodeMilestag2()` to be cleaner/better.
* Minor optimisations.

Note: No real-world data as yet.

crankyoldgit commented:
@vitos1k Can you please `git pull` your repo from github. I've add several commits. It should work now. Please test.

vitos1k commented:
> 
> 
> @vitos1k Can you please `git pull` your repo from github. I've add several commits. It should work now. Please test.

I've tested it and it still fails to detect miles tag. Through debugging i found out that in `_matchGeneric()` it fails to pass this row:
```cpp

 // Check if there is enough capture buffer to possibly have the message.
  if (remaining < min_remaining) return 0;  // Nope, so abort.
```

i've added this printouts to find out:

```cpp
// Check if there is enough capture buffer to possibly have the message.
  if (remaining < min_remaining) 
  {
    DPRINT("remaining:");
    DPRINT(remaining);
    DPRINT(" < min_remaining:");
    DPRINT(min_remaining);
    DPRINTLN("  return FAIL");
    return 0;
  }  // Nope, so abort.
```

and got this output in my terminal:

```
Attempting MilesTag2 decode
remaining:29 < min_remaining:50  return FAIL
remaining:29 < min_remaining:30  return FAIL
Attempting MilesTag2 decode
remaining:29 < min_remaining:50  return FAIL
remaining:29 < min_remaining:30  return FAIL
```

crankyoldgit commented:
> > @vitos1k Can you please `git pull` your repo from github. I've add several commits. It should work now. Please test.
> 
> I've tested it and it still fails to detect miles tag. Through debugging i found out that in `_matchGeneric()` it fails to pass this row:
> 
> ```c++
>  // Check if there is enough capture buffer to possibly have the message.
>   if (remaining < min_remaining) return 0;  // Nope, so abort.
> ```
> 
> i've added this printouts to find out:
> 
> ```c++
> // Check if there is enough capture buffer to possibly have the message.
>   if (remaining < min_remaining) 
>   {
>     DPRINT("remaining:");
>     DPRINT(remaining);
>     DPRINT(" < min_remaining:");
>     DPRINT(min_remaining);
>     DPRINTLN("  return FAIL");
>     return 0;
>   }  // Nope, so abort.
> ```
> 
> and got this output in my terminal:
> 
> ```
> Attempting MilesTag2 decode
> remaining:29 < min_remaining:50  return FAIL
> remaining:29 < min_remaining:30  return FAIL
> Attempting MilesTag2 decode
> remaining:29 < min_remaining:50  return FAIL
> remaining:29 < min_remaining:30  return FAIL
> ```

Got the raw message it's failing on so I can debug it?

vitos1k commented:
> 
> Got the raw message it's failing on so I can debug it?



sure 
```cpp
uint16_t rawData[29] = {2440, 602,  608, 600,  606, 600,  606, 600,  602, 606,  1208, 602,  1216, 596,  604, 600,  1214, 598,  1212, 600,  1208, 604,  1208, 602,  606, 600,  610, 596,  1210};  // MILESTAG2 379
```
btw if you call 
```cpp
irsend.sendMilestag2(0x0379);
```
It would produce the same raw data.

crankyoldgit commented:
> > Got the raw message it's failing on so I can debug it?
> 
> sure
> 
> ```c++
> uint16_t rawData[29] = {2440, 602,  608, 600,  606, 600,  606, 600,  602, 606,  1208, 602,  1216, 596,  604, 600,  1214, 598,  1212, 600,  1208, 604,  1208, 602,  606, 600,  610, 596,  1210};  // MILESTAG2 379
> ```
> 
> btw if you call
> 
> ```c++
> irsend.sendMilestag2(0x0379);
> ```
> 
> It would produce the same raw data.

Thanks. I've probably got an off-by-one error in my new code.

vitos1k commented:
> 
> 
> > > Got the raw message it's failing on so I can debug it?
> > 
> > 
> > sure
> > ```c++
> > uint16_t rawData[29] = {2440, 602,  608, 600,  606, 600,  606, 600,  602, 606,  1208, 602,  1216, 596,  604, 600,  1214, 598,  1212, 600,  1208, 604,  1208, 602,  606, 600,  610, 596,  1210};  // MILESTAG2 379
> > ```
> > 
> > 
> > btw if you call
> > ```c++
> > irsend.sendMilestag2(0x0379);
> > ```
> > 
> > 
> > It would produce the same raw data.
> 
> Thanks. I've probably got an off-by-one error in my new code.

looking through your `matchData()` function,  I was wondering, isn't it dangerous to call function inside function itself?
 Also if:
` (MSBfirst == false) & (expectlastspace == false) `
this would reverse bits twice, after matching all even bits, then adding one more bit, and then reversing it again, or am i missing something?

crankyoldgit committed:
Fix off-by-one error in calculating min_remaining.
* Needed to subtract `1` from `min_remaining` in `_matchGeneric()` when not expecting a trailing space.
* Unit tests to cover those cases.

crankyoldgit commented:
> looking through your `matchData()` function, I was wondering, isn't it dangerous to call function inside function itself?

No, not if you know what you're doing. It's call [recursion](https://en.wikipedia.org/wiki/Recursion_(computer_science)).

> Also if:
> `(MSBfirst == false) & (expectlastspace == false)`
> this would reverse bits twice, after matching all even bits, then adding one more bit, and then reversing it again, or am i missing something?

I think you're missing something.

I've fixed the off-by-one error, and pushed the commit to your repo, so please `git pull` again and test it out.

It should now match your example case. Let me know how it goes please?

vitos1k commented:
> 
> 
> > looking through your `matchData()` function, I was wondering, isn't it dangerous to call function inside function itself?
> 
> No, not if you know what you're doing. It's call [recursion](https://en.wikipedia.org/wiki/Recursion_(computer_science)).
> 
> > Also if:
> > `(MSBfirst == false) & (expectlastspace == false)`
> > this would reverse bits twice, after matching all even bits, then adding one more bit, and then reversing it again, or am i missing something?
> 
> I think you're missing something.
> 
> I've fixed the off-by-one error, and pushed the commit to your repo, so please `git pull` again and test it out.
> 
> It should now match your example case. Let me know how it goes please?

Yes! it works now as expected! Thanks!

crankyoldgit reviewed (commented):
Thanks for the confirmation. Looks good to me now.
I'll get @NiKiZe to also review this PR as I've contributed significantly too. i.e. To review my code ;-)

darshkpatel reviewed (approved):
LGTM

crankyoldgit commented:
Thanks @darshkpatel !

crankyoldgit commented:
FYI, the changes mentioned above have now been included in the new v2.7.15 release of the library.

vitos1k commented:
That's great to hear! Thanks!
